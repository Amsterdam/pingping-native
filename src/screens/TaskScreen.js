import React, {useState} from 'react';
import {View, StyleSheet, ScrollView, Image} from 'react-native';
import PropTypes from 'prop-types';
import {Content, Container} from 'native-base';
import YouTube from 'react-native-youtube';
import HTML from 'react-native-render-html';
import {useMutation, useQuery} from '@apollo/client';
import GET_ROUTE_QUERY from '../apollo/Query/getRoute';
import COMPLETE_TASK_MUTATION from '../apollo/Mutation/completeTaskMutation';
import LabeledHeader from '../components/header/LabeledHeader';
import Title from '../components/typography/Title';
import Button from '../components/OnboardingButton';
import {appColors} from '../config/colors';
import WebViewModal from '../components/modals/WebViewModal';
import Loading from '../components/LoadingComponent';

const TaskScreen = ({navigation, route}) => {
  const {task, routeId} = route.params;
  const [completeTask] = useMutation(COMPLETE_TASK_MUTATION);
  const {refetch} = useQuery(GET_ROUTE_QUERY, {
    variables: {routeId},
  });
  const [urlToVisit, setUrlToVisit] = useState('https://amsterdam.nl');
  const [webViewOpen, setWebviewOpen] = useState(false);
  const [loading, setLoading] = useState(false);

  const doCompleteTask = async () => {
    setLoading(true);
    try {
      await completeTask({
        variables: {taskId: task.taskId},
      });

      /* This code checks if the route is done in order to show the celebration modal yes/no */
      const routeResponse = await refetch();
      const routeDone =
        routeResponse?.data?.getRoute?.tasks.filter(
          (task) => task.status !== 'Completed',
        ).length === 0;
      routeDone &&
        navigation.navigate('CityPings', {
          screen: 'CityPingsModalScreen',
          params: {pings: routeResponse.data.getRoute.totalPoints},
          initial: false,
        });
      setLoading(false);
      navigation.goBack();
    } catch (error) {
      setLoading(false);
      console.log(error);
    }
  };

  const renderMedia = (media) => {
    switch (media.type) {
      case 'YouTube':
        return (
          <YouTube
            videoId={task.media.value} // The YouTube video ID
            play={false} // control playback of video with true/false
            loop={false} // control whether the video should loop when ended
            apiKey="AIzaSyBqyTzXy2qEPvGLXDxZ4En_rP6krgVvtFk"
            style={styles.videoContainer}
          />
        );
      case 'Image':
        return (
          <Image
            source={{uri: task.media.value}}
            style={styles.imageContainer}
          />
        );
      default:
        break;
    }
  };

  const linkPressed = (event, href) => {
    setUrlToVisit(href);
    setWebviewOpen(true);
  };

  const closeModal = () => {
    setWebviewOpen(false);
  };

  const taskStatus = task.status === 'Completed';

  return (
    <React.Fragment>
      <Container>
        <LabeledHeader
          filledHeader
          navigation={navigation}
          title={task.headerTitle}
        />
        <Content contentContainerStyle={styles.contentContainer}>
          {task?.media && renderMedia(task.media)}

          <View style={styles.textContainer}>
            <ScrollView>
              <Title>{task.title}</Title>
              <View style={styles.descriptionContainer}>
                <HTML
                  html={task.description}
                  baseFontStyle={styles.htmlFontStyle}
                  onLinkPress={(event, href) => {
                    linkPressed(event, href);
                  }}
                />
              </View>
            </ScrollView>
          </View>
        </Content>
        {taskStatus && (
          <Title style={styles.completedTagLine} align="center">
            Je {task.headerTitle} is gefikst
          </Title>
        )}
        <View style={styles.buttonContainer}>
          {taskStatus ? (
            <Button
              style={styles.buttonHelp}
              label="terug"
              onPress={() => navigation.goBack()}
            />
          ) : (
            <React.Fragment>
              <Button style={styles.buttonHelp} label="Hulp nodig?" />
              <Button label="Gelukt!" onPress={doCompleteTask} />
            </React.Fragment>
          )}
        </View>
        {loading && <Loading />}
      </Container>

      <WebViewModal
        urlToVisit={urlToVisit}
        closeModal={closeModal}
        webViewOpen={webViewOpen}
        setWebviewOpen={setWebviewOpen}
      />
    </React.Fragment>
  );
};

const styles = StyleSheet.create({
  contentContainer: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: '#fff',
  },
  videoContainer: {
    alignSelf: 'stretch',
    height: 200,
  },
  imageContainer: {
    alignSelf: 'stretch',
    height: 200,
  },
  textContainer: {
    flex: 1,
    paddingHorizontal: 30,
    paddingTop: 20,
  },
  descriptionContainer: {
    justifyContent: 'space-around',
    flex: 1,
  },
  buttonContainer: {
    paddingHorizontal: 40,
    marginVertical: 10,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  buttonHelp: {
    backgroundColor: appColors.greyedOut,
    flex: 1,
    marginRight: 10,
    justifyContent: 'center',
  },
  htmlFontStyle: {
    fontFamily: 'Raleway',
    fontSize: 15,
    lineHeight: 25,
  },
  completedTagLine: {fontSize: 14, color: appColors.primary, marginTop: 10},
});

TaskScreen.propTypes = {
  navigation: PropTypes.object.isRequired,
  route: PropTypes.object.isRequired,
};

export default TaskScreen;
