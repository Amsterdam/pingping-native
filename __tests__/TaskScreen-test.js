import React from 'react';

import {MockedProvider} from '@apollo/client/testing';
import renderer from 'react-test-renderer';

import TaskScreen from '../src/screens/TaskScreen';

const task = {title: 'A title', description: 'a description'};

jest.useFakeTimers();
const navigation = {navigate: jest.fn()};

test('renders correctly', () => {
	const tree = renderer
		.create(
			<MockedProvider mocks={[]}>
				<TaskScreen navigation={navigation} route={{params: {task}}} />
			</MockedProvider>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
