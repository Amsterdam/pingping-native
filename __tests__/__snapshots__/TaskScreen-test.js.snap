// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "height": 1334,
      }
    }
  >
    <View>
      <RCTSafeAreaView
        emulateUnlessSupported={true}
        style={
          Object {
            "backgroundColor": "#0d2036",
          }
        }
      >
        <View
          noShadow={true}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderBottomColor": "transparent",
                "borderBottomWidth": 0.5,
                "elevation": 0,
                "flexDirection": "row",
                "height": 64,
                "justifyContent": "center",
                "left": 0,
                "paddingLeft": 6,
                "paddingRight": 10,
                "paddingTop": 18,
                "right": 0,
                "shadowColor": null,
                "shadowOffset": null,
                "shadowOpacity": null,
                "shadowRadius": null,
                "top": 0,
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#0d2036",
              },
            ]
          }
          transparent={true}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-start",
                  "alignSelf": "center",
                  "flex": 1,
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": null,
                  "backgroundColor": "transparent",
                  "borderBottomWidth": null,
                  "borderColor": null,
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 0,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "marginLeft": -3,
                  "opacity": 1,
                  "paddingBottom": 6,
                  "paddingLeft": 4,
                  "paddingRight": 6,
                  "paddingTop": 6,
                  "shadowColor": null,
                  "shadowOffset": null,
                  "shadowOpacity": null,
                  "shadowRadius": null,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    Array [
                      Object {
                        "color": "#000",
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#007aff",
                        "fontSize": 34,
                        "marginLeft": 1,
                        "marginRight": 2,
                        "marginTop": 0,
                        "paddingTop": 1,
                      },
                      Object {
                        "color": "#fff",
                      },
                    ],
                    Object {
                      "fontFamily": "anticon",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                type="AntDesign"
              >
                î˜œ
              </Text>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 14,
                  "marginTop": 5,
                  "textAlign": "left",
                },
              ]
            }
            uppercase={false}
          >
            none
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-end",
                  "alignSelf": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          />
        </View>
      </RCTSafeAreaView>
    </View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={
          Array [
            Object {
              "padding": undefined,
            },
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "flex": 1,
            },
          ]
        }
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        enableAutomaticScroll={true}
        enableOnAndroid={false}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardShouldPersistTaps="handled"
        keyboardSpace={0}
        onScroll={[Function]}
        resetKeyboardSpace={[Function]}
        resetScrollToCoords={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollIntoView={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "flex": 1,
          }
        }
        update={[Function]}
        viewIsInsideTabBar={false}
      >
        <View>
          <RCTYouTube
            loopProp={true}
            onChangeFullscreen={[Function]}
            onChangeQuality={[Function]}
            onChangeState={[Function]}
            onError={[Function]}
            onProgress={[Function]}
            onReady={[Function]}
            play={false}
            playerParams={
              Object {
                "playerVars": Object {
                  "controls": undefined,
                  "fs": undefined,
                  "loop": 1,
                  "modestbranding": undefined,
                  "origin": undefined,
                  "playlist": "ZG0vCfivpQ8",
                  "playsinline": 1,
                  "rel": undefined,
                  "showinfo": undefined,
                },
                "playlistId": undefined,
                "videoId": "ZG0vCfivpQ8",
              }
            }
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "alignSelf": "stretch",
                  "height": 300,
                },
              ]
            }
            videoId="ZG0vCfivpQ8"
          />
          <View
            style={
              Object {
                "flex": 1,
                "padding": 30,
                "position": "relative",
                "width": 750,
              }
            }
          >
            <RCTScrollView>
              <View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      },
                      Object {
                        "fontFamily": "Heavitas",
                        "fontSize": 30,
                        "textAlign": "left",
                      },
                    ]
                  }
                  uppercase={false}
                >
                  A title
                </Text>
                <View
                  style={
                    Object {
                      "flex": 1,
                      "justifyContent": "space-around",
                    }
                  }
                >
                  <View
                    style={Object {}}
                  >
                    <Text
                      allowFontScaling={true}
                      selectable={false}
                      style={
                        Array [
                          Object {},
                          Object {},
                        ]
                      }
                    >
                      <Text
                        allowFontScaling={true}
                        style={
                          Object {
                            "fontFamily": "Raleway",
                            "fontSize": 15,
                            "lineHeight": 25,
                          }
                        }
                      >
                        a description
                      </Text>
                    </Text>
                  </View>
                </View>
              </View>
            </RCTScrollView>
          </View>
        </View>
      </RCTScrollView>
    </RCTSafeAreaView>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 20,
          "paddingHorizontal": 40,
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        forwardedRef={[Function]}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#b9b9b9",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 10,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "borderWidth": undefined,
            "elevation": 2,
            "flex": 1,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "center",
            "marginRight": 10,
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "#fff",
                  "fontFamily": "System",
                  "fontSize": 16.5,
                  "marginLeft": 0,
                  "marginRight": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 14,
                  "paddingTop": 3,
                },
              ]
            }
            uppercase={false}
          >
            Hulp nodig?
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        forwardedRef={[Function]}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fb9f00",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 10,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "borderWidth": undefined,
            "elevation": 2,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "#fff",
                  "fontFamily": "System",
                  "fontSize": 16.5,
                  "marginLeft": 0,
                  "marginRight": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 14,
                  "paddingTop": 3,
                },
              ]
            }
            uppercase={false}
          >
            Gelukt!
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(1,1,1,0.8)",
          "borderRadius": 35,
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 35,
            "flex": 1,
            "marginTop": 75,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fb9f00",
              "borderTopLeftRadius": 35,
              "borderTopRightRadius": 35,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 30,
                  "marginTop": 10,
                  "textAlign": "left",
                },
              ]
            }
            uppercase={false}
          >
            SLUITEN
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "padding": 5,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "overflow": "hidden",
                },
                undefined,
              ]
            }
          >
            <RNCWebView
              cacheEnabled={true}
              injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
              injectedJavaScriptForMainFrameOnly={true}
              javaScriptEnabled={true}
              messagingEnabled={false}
              onContentProcessDidTerminate={[Function]}
              onHttpError={[Function]}
              onLoadingError={[Function]}
              onLoadingFinish={[Function]}
              onLoadingProgress={[Function]}
              onLoadingStart={[Function]}
              onMessage={[Function]}
              onShouldStartLoadWithRequest={[Function]}
              source={
                Object {
                  "uri": "https://amsterdam.nl",
                }
              }
              style={
                Array [
                  Object {
                    "flex": 1,
                    "overflow": "hidden",
                  },
                  Object {
                    "backgroundColor": "#ffffff",
                  },
                  undefined,
                ]
              }
              useSharedProcessPool={true}
            />
          </View>
        </View>
      </View>
    </View>
  </Modal>,
]
`;
